google-cloud-bigquery/lib/google/cloud/bigquery/dataset.rb

```diff
@@ -2056,6 +2067,15 @@ def ensure_gapi!
           reload!
         end
 
+        ##
+        # Fetch gapi and memoize whether resource exists.
+        def gapi_exists?
+          reload!
+          @exists = true
```

Comment:
This method should also set `@deleted = nil`, or `#reload!` should set `@deleted = nil` before the API call is made.

Or, we could move the newly added logic from `#reload!` to this private method, and possibly rename it `#safe_reload!`.